aws/resource_aws_guardduty_detector.go

```diff
@@ -0,0 +1,111 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/guardduty"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceAwsGuardDutyDetector() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsGuardDutyDetectorCreate,
+		Read:   resourceAwsGuardDutyDetectorRead,
+		Update: resourceAwsGuardDutyDetectorUpdate,
+		Delete: resourceAwsGuardDutyDetectorDelete,
+
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"enable": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  true,
+			},
+			"account_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceAwsGuardDutyDetectorCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).guarddutyconn
+
+	input := guardduty.CreateDetectorInput{
+		Enable: aws.Bool(d.Get("enable").(bool)),
+	}
+
+	log.Printf("[DEBUG] Creating GuardDuty Detector: %#v", input)
+	output, err := conn.CreateDetector(&input)
+	if err != nil {
+		return fmt.Errorf("Creating GuardDuty Detector failed: %s", err.Error())
+	}
+	d.SetId(*output.DetectorId)
+
+	return resourceAwsGuardDutyDetectorRead(d, meta)
+}
+
+func resourceAwsGuardDutyDetectorRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).guarddutyconn
+	input := guardduty.GetDetectorInput{
+		DetectorId: aws.String(d.Id()),
+	}
+
+	log.Printf("[DEBUG] Reading GuardDuty Detector: %#v", input)
+	gdo, err := conn.GetDetector(&input)
+	if err != nil {
+		if isAWSErr(err, "NoSuchEntityException", "The request is rejected because the input detectorId is not owned by the current account.") {
+			log.Printf("[WARN] GuardDuty detector %q not found, removing from state", d.Id())
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("Reading GuardDuty Detector '%s' failed: %s", d.Id(), err.Error())
+	}
+
+	d.Set("account_id", meta.(*AWSClient).accountid)
```

Comment:
I believe it makes more sense to be able to link to the exact resource you're trying to get the account ID of (or region in other use cases), rather than requiring and depending on a potentially misconfigured data source in a multi-account configuration. Implementing it here is a purely UX aid in this resource to reduce additional configuration/documentation/knowledge for multi-account setups. When I first started using Terraform it was confusing at first to learn that `aws_caller_identity` (nothing in its naming implies it has an AWS account ID) existed to properly setup these types of configurations. 

As I mentioned in the original PR comment, it'd be nice if the provider could add account ID and region as automatic attributes for all resources to not require and reduce room for user error with `aws_caller_identity`/`aws_region`.

If you'd like this removed, let me know.

Reply:
That's reasonable explanation, I just wonder which resource would require the account ID associated with the detector? i.e. can you provide an example of where would you reference this attribute?

I'm ok with keeping it there if we have a real use-case for it. If we don't then I'd rather remove it.